### PR TITLE
SW-694 Dynamic product groups categories must work as cross-selling category

### PR DESCRIPTION
**Pull request by TobiasGraml11**

### CODE REVIEW

File: src/Export/XmlExport.php [hunk 1]
```diff
@@ -130,16 +130,48 @@ private function exportSingleItem(
         return $xmlProduct->getXmlItem();
     }
 
-    private function isProductInCrossSellingCategory(ProductEntity $productEntity): bool
+    private function isProductInCrossSellingCategory(ProductEntity $productEntity): ?CategoryEntity
```
Comment (dblaichinger) [inline src/Export/XmlExport.php:133]: If method name starts with `is` I would expect a boolean return, maybe `getEntityFromCrossSellingCategories`?

Comment (dblaichinger) [inline src/Export/XmlExport.php]: If method name starts with `is` I would expect a boolean return, maybe `getEntityFromCrossSellingCategories`?

File: src/Export/XmlExport.php [hunk 2]
```diff
@@ -130,16 +130,48 @@ private function exportSingleItem(
...
-            $categoryId = $category ? $category->getId() : null;
+        if (count($this->crossSellingCategories)) {
+            $categories = array_merge(
+                $this->getAssignedCategories($productEntity),
+                $this->getDynamicProductGroupCategories($productEntity)
+            );
+
+            foreach ($categories as $categoryId => $category) {
+                if (in_array($categoryId, $this->crossSellingCategories)) {
+                    return $category;
```
Comment (dblaichinger) [inline src/Export/XmlExport.php:143]: This would only return first match, is this fine?

File: src/Export/XmlExport.php [hunk 3]
```diff
@@ -130,16 +130,48 @@ private function exportSingleItem(
...
+    private function getDynamicProductGroupCategories(ProductEntity $productEntity): array
+    {
+        $dynamicProductGroupService = null;
+        if ($this->container->has('fin_search.dynamic_product_group')) {
+            $dynamicProductGroupService = $this->container->get('fin_search.dynamic_product_group');
+        }
 
-            return (in_array($categoryId, $this->crossSellingCategories));
+        if ($dynamicProductGroupService) {
+            return $dynamicProductGroupService->getCategories($productEntity->getId());
```
Comment (dblaichinger) [inline src/Export/XmlExport.php]: This could be moved into the upper if.

File: tests/Traits/DataHelpers/CategoryHelper.php [hunk 4]
```diff
@@ -27,9 +28,17 @@ public function createTestCategory(array $categories)
         /** @var CategoryCollection $result */
         $result = $repository->search($criteria, Context::createDefaultContext());
 
-        /** @var CategoryEntity $first */
-        $first = $result->first();
+        return $result->first();
+    }
+
+    public function createBasicCategory(?array $overrideData = []): CategoryEntity
```
Comment (dblaichinger) [inline tests/Traits/DataHelpers/CategoryHelper.php]: As discussed, please try to combine into `createTestCategory`

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: src/Export/XmlExport.php (hunk 2 above)
Comment (TobiasGraml11) [reply src/Export/XmlExport.php:143]: I would say so, yes.
If needed, I can change it to listing all categories.

**Review by dblaichinger — APPROVED** [review]

File: tests/Export/ProductIdExportTest.php [hunk 5]
```diff
@@ -6,17 +6,14 @@
...
 use FINDOLOGIC\FinSearch\Export\ProductIdExport;
 use FINDOLOGIC\FinSearch\Export\XmlExport;
+use Shopware\Core\Content\Category\CategoryEntity;
+use Shopware\Core\Content\Product\ProductEntity;
 use Shopware\Storefront\Framework\Routing\Router;
 
 class ProductIdExportTest extends XmlExportTest
 {
-    public function testProductsInCrossSellCategoriesAreNotWrappedAndErrorIsLogged(): void
+    public function buildItemsAndAssertError(ProductEntity $product, CategoryEntity $category)
```
Comment (TheKeymaster) [inline tests/Export/ProductIdExportTest.php:15]: This method is never called, therefore nothing tests this logic for the product id export.

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: tests/Export/ProductIdExportTest.php (hunk 5 above)
Comment (TobiasGraml11) [reply tests/Export/ProductIdExportTest.php:15]: The test class extends `XmlExportTest`, so this is just an extended version of the helper function defined there.
It is called during the tests of `XmlExportTest`.

image

**Review by TheKeymaster — APPROVED** [review]

**Pull request merged**